File: sdks/javascript/src/engine/time.ts
/**
 * Time helpers — millisecond timestamp + formatted local time string.
 */

export function now(): number {
  return Date.now();
}

export function formatTs(ts = now()): string {
  const d = new Date(ts);
  const h = String(d.getHours()).padStart(2, '0');
  const m = String(d.getMinutes()).padStart(2, '0');
  const s = String(d.getSeconds()).padStart(2, '0');
  const ms = String(d.getMilliseconds()).padStart(3, '0');
  return `[${h}:${m}:${s}.${ms}]`;
}

/** Generate a traceId with the format `<scope>-<timestamp6>-<rand3>`. */
export function makeTraceId(scope: string): string {
  const ts = String(now()).slice(-6);
  const rand = Math.random().toString(36).slice(2, 5);
  return `${scope}-${ts}-${rand}`;
}

/** Generate a spanId like `span-3`. */
let spanCounter = 0;
export function makeSpanId(): string {
  return `span-${++spanCounter}`;
}

File: sdks/python/src/tracelink/engine/time.py
"""Time and ID helpers."""

from __future__ import annotations

import os
import time
from datetime import datetime


def now_ms() -> int:
    return int(time.time() * 1000)


def monotonic_ms() -> int:
    """Return a monotonic millisecond clock for elapsed-time measurements."""
    return time.perf_counter_ns() // 1_000_000


def format_ts(ts: int | None = None) -> str:
    if ts is None:
        ts = now_ms()
    dt = datetime.fromtimestamp(ts / 1000)
    return f"[{dt.strftime('%H:%M:%S')}.{dt.microsecond // 1000:03d}]"


def make_trace_id(scope: str) -> str:
    """Generate traceId with format `<scope>-<timestamp6>-<rand3>`."""
    ts_suffix = str(now_ms())[-6:]
    rand_suffix = os.urandom(2).hex()[:3]
    return f"{scope}-{ts_suffix}-{rand_suffix}"


def make_span_id(counter: int) -> str:
    return f"span-{counter}"
